go/test/endtoend/vreplication/materialize_test.go

```diff
@@ -96,3 +96,144 @@ func TestShardedMaterialize(t *testing.T) {
 		`[[INT64(3) VARBINARY("def")] [INT64(5) VARBINARY("def")]]`)
 }
+
+/*
+ * The following sections are related to TestMaterialize. TestMaterialize is intended to test some edge cases. Currently
+ * it tests
+ * - the case where the same column is referred to multiple times
+ * - use of mysql functions in the filter
+ * - use of a custom function in the filter
+ */
+
+const smMaterializeSchemaSource = `
+	CREATE TABLE mat (
+	id bigint NOT NULL,
+	val varbinary(10) NOT NULL,
+	ts timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
+	PRIMARY KEY (id)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
+`
+
+const smMaterializeVSchemaSource = `
+{
+  "sharded": true,
+  "tables": {
+      "mat": {
+          "column_vindexes": [
+              {
+                  "column": "id",
+                  "name": "hash"
+              }
+          ]
+      }
+  },
+  "vindexes": {
+      "hash": {
+          "type": "hash"
+      }
+  }
+}
+`
+const smMaterializeSchemaTarget = `
+	CREATE TABLE mat2 (
+	id bigint NOT NULL,
+	val varbinary(10) NOT NULL,
+	ts timestamp NOT NULL,
+	day int NOT NULL,
+	month int NOT NULL,
+    x int not null,
+	PRIMARY KEY (id)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
+`
+
+const smMaterializeVSchemaTarget = `
+{
+  "sharded": true,
+  "tables": {
+      "mat2": {
+          "column_vindexes": [
+              {
+                  "column": "id",
+                  "name": "hash"
+              }
+          ]
+      }
+  },
+  "vindexes": {
+      "hash": {
+          "type": "hash"
+      }
+  }
+}
+`
+const smMaterializeSpec2 = `{"workflow": "wf1", "source_keyspace": "source", "target_keyspace": "target", "table_settings": [ {"target_table": "mat2", "source_expression": "select id, val, ts, dayofmonth(ts) as day, month(ts) as month, custom1(id, val) as x from mat"  }] }`
+
+const materializeInitDataQuery = `insert into mat(id, val, ts) values (1, 'abc', '2021-10-9 16:17:36'), (2, 'def', '2021-11-10 16:17:36')`
+
+const customFunc = `
+CREATE FUNCTION custom1(id int, val varbinary(10))
+RETURNS int
+DETERMINISTIC
+RETURN id * length(val);
+`
+
+// TestMaterialize: details mentioned above
+func testMaterialize(t *testing.T) {
+	defaultCellName := "zone1"
+	allCells := []string{"zone1"}
+	allCellNames = "zone1"
+	vc = NewVitessCluster(t, "TestMaterialize", allCells, mainClusterConfig)
+	sourceKs := "source"
+	targetKs := "target"
+	require.NotNil(t, vc)
+	defaultReplicas = 0 // because of CI resource constraints we can only run this test with primary tablets
+	defer func() { defaultReplicas = 1 }()
+
+	defer vc.TearDown(t)
+
+	defaultCell = vc.Cells[defaultCellName]
+	vc.AddKeyspace(t, []*Cell{defaultCell}, sourceKs, "-", smMaterializeVSchemaSource, smMaterializeSchemaSource, defaultReplicas, defaultRdonly, 300)
+	vtgate = defaultCell.Vtgates[0]
+	require.NotNil(t, vtgate)
+	vtgate.WaitForStatusOfTabletInShard(fmt.Sprintf("%s.%s.primary", sourceKs, "0"), 1)
```

Comment:
They seem to be synonyms but is there any reason to use "-" above and "0" here?

Reply:
We had a regression (a long while ago) where '-' and '0' were giving different results which wasn't caught by the tests at that time! So as a reflex  I have used the two interchangeably at times, but it is not strictly necessary for this test.